### ISSUE TITLE
cargo audit found 3 vulnerabilities

### ISSUE THREAD

**Issue by cataggar** [Azure/azure-sdk-for-rust#559]
```
PS C:\Users\cataggar\io\azure-sdk-for-rust> cargo audit
    Fetching advisory database from `https://github.com/RustSec/advisory-db.git`
      Loaded 384 security advisories (from C:\Users\cataggar\.cargo\advisory-db)
    Updating crates.io index
    Scanning Cargo.lock for vulnerabilities (193 crate dependencies)
Crate:         chrono
Version:       0.4.19
Title:         Potential segfault in `localtime_r` invocations
Date:          2020-11-10
ID:            RUSTSEC-2020-0159
URL:           https://rustsec.org/advisories/RUSTSEC-2020-0159
Solution:      No safe upgrade is available!
Dependency tree:
chrono 0.4.19
├── oauth2 4.1.0
│   ├── azure_security_keyvault 0.1.0
│   ├── azure_identity 0.1.0
│   │   ├── azure_storage_datalake 0.1.0
│   │   ├── azure_storage_blobs 0.1.0
│   │   │   └── azure_identity 0.1.0
│   │   ├── azure_storage 0.1.0
│   │   │   ├── azure_storage_queues 0.1.0
│   │   │   ├── azure_storage_datalake 0.1.0
│   │   │   ├── azure_storage_blobs 0.1.0
│   │   │   ├── azure_identity 0.1.0
│   │   │   └── azure_data_tables 0.1.0
│   │   └── azure_security_keyvault 0.1.0
│   └── azure_core 0.1.0
│       ├── iot_hub 0.1.0
│       ├── azure_storage_queues 0.1.0
│       ├── azure_storage_datalake 0.1.0
│       ├── azure_storage_blobs 0.1.0
│       ├── azure_storage 0.1.0
│       ├── azure_security_keyvault 0.1.0
│       ├── azure_messaging_servicebus 0.1.0
│       ├── azure_messaging_eventgrid 0.1.0
│       ├── azure_identity 0.1.0
│       ├── azure_data_tables 0.1.0
│       └── azure_cosmos 0.1.0
├── iot_hub 0.1.0
├── azure_storage_queues 0.1.0
├── azure_storage_datalake 0.1.0
├── azure_storage_blobs 0.1.0
├── azure_storage 0.1.0
├── azure_security_keyvault 0.1.0
├── azure_messaging_servicebus 0.1.0
├── azure_messaging_eventgrid 0.1.0
├── azure_identity 0.1.0
├── azure_data_tables 0.1.0
├── azure_cosmos 0.1.0
└── azure_core 0.1.0

Crate:         time
Version:       0.1.43
Title:         Potential segfault in the time crate
Date:          2020-11-18
ID:            RUSTSEC-2020-0071
URL:           https://rustsec.org/advisories/RUSTSEC-2020-0071
Solution:      Upgrade to >=0.2.23
Dependency tree:
time 0.1.43
└── chrono 0.4.19
    ├── oauth2 4.1.0
    │   ├── azure_security_keyvault 0.1.0
    │   ├── azure_identity 0.1.0
    │   │   ├── azure_storage_datalake 0.1.0
    │   │   ├── azure_storage_blobs 0.1.0
    │   │   │   └── azure_identity 0.1.0
    │   │   ├── azure_storage 0.1.0
    │   │   │   ├── azure_storage_queues 0.1.0
    │   │   │   ├── azure_storage_datalake 0.1.0
    │   │   │   ├── azure_storage_blobs 0.1.0
    │   │   │   ├── azure_identity 0.1.0
    │   │   │   └── azure_data_tables 0.1.0
    │   │   └── azure_security_keyvault 0.1.0
    │   └── azure_core 0.1.0
    │       ├── iot_hub 0.1.0
    │       ├── azure_storage_queues 0.1.0
    │       ├── azure_storage_datalake 0.1.0
    │       ├── azure_storage_blobs 0.1.0
    │       ├── azure_storage 0.1.0
    │       ├── azure_security_keyvault 0.1.0
    │       ├── azure_messaging_servicebus 0.1.0
    │       ├── azure_messaging_eventgrid 0.1.0
    │       ├── azure_identity 0.1.0
    │       ├── azure_data_tables 0.1.0
    │       └── azure_cosmos 0.1.0
    ├── iot_hub 0.1.0
    ├── azure_storage_queues 0.1.0
    ├── azure_storage_datalake 0.1.0
    ├── azure_storage_blobs 0.1.0
    ├── azure_storage 0.1.0
    ├── azure_security_keyvault 0.1.0
    ├── azure_messaging_servicebus 0.1.0
    ├── azure_messaging_eventgrid 0.1.0
    ├── azure_identity 0.1.0
    ├── azure_data_tables 0.1.0
    ├── azure_cosmos 0.1.0
    └── azure_core 0.1.0

Crate:         tokio
Version:       1.13.0
Title:         Data race when sending and receiving after closing a `oneshot` channel
Date:          2021-11-16
ID:            RUSTSEC-2021-0124
URL:           https://rustsec.org/advisories/RUSTSEC-2021-0124
Solution:      Upgrade to >=1.8.4, <1.9.0 OR >=1.13.1
Dependency tree:
tokio 1.13.0
├── tokio-util 0.6.7
│   └── h2 0.3.3
│       └── hyper 0.14.10
│           ├── reqwest 0.11.4
│           │   ├── oauth2 4.1.0
│           │   │   ├── azure_security_keyvault 0.1.0
│           │   │   ├── azure_identity 0.1.0
│           │   │   │   ├── azure_storage_datalake 0.1.0
│           │   │   │   ├── azure_storage_blobs 0.1.0
│           │   │   │   │   └── azure_identity 0.1.0
│           │   │   │   ├── azure_storage 0.1.0
│           │   │   │   │   ├── azure_storage_queues 0.1.0
│           │   │   │   │   ├── azure_storage_datalake 0.1.0
│           │   │   │   │   ├── azure_storage_blobs 0.1.0
│           │   │   │   │   ├── azure_identity 0.1.0
│           │   │   │   │   └── azure_data_tables 0.1.0
│           │   │   │   └── azure_security_keyvault 0.1.0
│           │   │   └── azure_core 0.1.0
│           │   │       ├── iot_hub 0.1.0
│           │   │       ├── azure_storage_queues 0.1.0
│           │   │       ├── azure_storage_datalake 0.1.0
│           │   │       ├── azure_storage_blobs 0.1.0
│           │   │       ├── azure_storage 0.1.0
│           │   │       ├── azure_security_keyvault 0.1.0
│           │   │       ├── azure_messaging_servicebus 0.1.0
│           │   │       ├── azure_messaging_eventgrid 0.1.0
│           │   │       ├── azure_identity 0.1.0
│           │   │       ├── azure_data_tables 0.1.0
│           │   │       └── azure_cosmos 0.1.0
│           │   ├── iot_hub 0.1.0
│           │   ├── azure_storage_blobs 0.1.0
│           │   ├── azure_storage 0.1.0
│           │   ├── azure_security_keyvault 0.1.0
│           │   ├── azure_identity 0.1.0
│           │   ├── azure_cosmos 0.1.0
│           │   └── azure_core 0.1.0
│           ├── iot_hub 0.1.0
│           ├── hyper-tls 0.5.0
│           │   └── reqwest 0.11.4
│           ├── hyper-rustls 0.22.1
│           │   ├── reqwest 0.11.4
│           │   ├── iot_hub 0.1.0
│           │   ├── azure_cosmos 0.1.0
│           │   └── azure_core 0.1.0
│           ├── azure_cosmos 0.1.0
│           └── azure_core 0.1.0
├── tokio-rustls 0.22.0
│   ├── reqwest 0.11.4
│   └── hyper-rustls 0.22.1
├── tokio-native-tls 0.3.0
│   ├── reqwest 0.11.4
│   └── hyper-tls 0.5.0
├── stop-token 0.6.1
│   └── azure_cosmos 0.1.0
├── reqwest 0.11.4
├── iot_hub 0.1.0
├── hyper-tls 0.5.0
├── hyper-rustls 0.22.1
├── hyper 0.14.10
├── h2 0.3.3
├── azure_storage_queues 0.1.0
├── azure_storage_datalake 0.1.0
├── azure_storage_blobs 0.1.0
├── azure_storage 0.1.0
├── azure_security_keyvault 0.1.0
├── azure_messaging_servicebus 0.1.0
├── azure_messaging_eventgrid 0.1.0
├── azure_identity 0.1.0
├── azure_data_tables 0.1.0
├── azure_cosmos 0.1.0
└── azure_core 0.1.0

Crate:         difference
Version:       2.0.0
Warning:       unmaintained
Title:         difference is unmaintained
Date:          2020-12-20
ID:            RUSTSEC-2020-0095
URL:           https://rustsec.org/advisories/RUSTSEC-2020-0095
Dependency tree:
difference 2.0.0
└── mockito 0.30.0
    └── azure_security_keyvault 0.1.0

error: 3 vulnerabilities found!
```

**Comment by cataggar** [comment]
oauth2 [says](https://github.com/ramosbugs/oauth2-rs/blob/30c934a87fa29ffb4780f545b1236d4ecb08c0e6/.github/workflows/main.yml#L72-L79) the thread safety issue does not apply since it only uses UTC.

```
   - name: Audit
        if: ${{ matrix.rust == 'stable' }}
        run: |
          cargo install --force cargo-audit
          # The chrono thread safety issue doesn't affect this crate since the crate does not rely
          # on the system's local time zone, only UTC. See:
          # https://github.com/chronotope/chrono/issues/499#issuecomment-946388161
          cargo audit --ignore RUSTSEC-2020-0159
```

**Comment by dbanty** [comment]
Is there a particular motivation for using `chrono` vs using `time` directly? I'm just curious, as I've found recently that I've been able to replace my usage of `chrono` with `time` to avoid those audit results (and have slightly fewer dependencies).

**Comment by cataggar** [comment]
I am open to using `time` directly. I noticed other libraries making that change too.

**Comment by dbanty** [comment]
I started a prototype of this to see how big of a lift it would be and encountered a slight problem in that some APIs assume formatting will always succeed, but the `.format` function on `time::OffsetDateTime` returns a `Result`. It's unclear to me how/when/why that function would fail in practice with a well-known format (`Rfc2822`) but it does mean that either the interface for `impl AddAsHeader for IfSourceModifiedSinceCondition` would have to change (which seems to be intended at some point anyway given the `add_as_header2` function) or we'd need to add the risk of panic with a `.unwrap` or `.expect`.

FWIW I did open https://github.com/time-rs/time/issues/444 for `time` requesting some insight into why/if an error can actually occur there.
